File: impls/perl/step3_env.pl
#!perl
use strict;
use warnings;
use FindBin;
use lib $FindBin::Bin;
use Term::ReadLine;
use Reader;
use Printer;
use Env;

my $repl_env=Env->new(
    '+'=>sub { my $val=0; $val+=shift while @_; return $val; },
    '-'=>sub { my $val=shift; $val-=shift while @_; return $val; },
    '*'=>sub { my $val=1; $val*=shift while @_; return $val; },
    '/'=>sub { my $val=shift; $val=int($val/shift) while @_; return $val; },
);
my %special=(
    'def!'=>\&eval_def_bang,
    'let*'=>\&eval_let_star,
);

sub READ {
    my $str=shift;
    return Reader::read_str($str);
}
sub EVAL {
    my ($val, $env)=@_;
    return eval_ast($val, $env) unless ref($val) eq 'MalList' && @$val;
    my $fn=$val->[0];
    if(ref $fn eq 'MalSymbol' && exists $special{$$fn}) {
        return $special{$$fn}->($val, $env)
    }
    return $special{$val->[0]}->($val, $env) if exists $special{$val->[0]};
    my $list=eval_ast($val, $env);
    $fn=shift @$list;
    return $fn->(@$list) if ref($fn) eq 'CODE';
    die "bad function\n";
}
sub PRINT {
    my $val=shift;
    return Printer::pr_str($val);
}

sub rep {
    my $line=shift;
    return PRINT(EVAL(READ($line), $repl_env));
}

sub eval_ast {
    my ($ast, $env)=@_;
    if(ref($ast) eq 'MalSymbol') {
        return $env->get($$ast);
    } elsif(ref($ast) eq 'MalList' || ref($ast) eq 'MalVector') {
        return bless [ map { EVAL($_, $env) } @$ast ], ref($ast);
    } elsif(ref($ast) eq 'MalHash') {
        return bless { map { EVAL($_, $env) } %$ast }, ref($ast);
    } else {
        return $ast;
    }
}

sub eval_def_bang {
    my ($ast, $env)=@_;
    my $key=$ast->[1];
    my $value=EVAL($ast->[2], $env);
    $env->set($key=>$value);
    return $value;
}
sub eval_let_star {
    my ($ast, $env)=@_;
    my $bindings=$ast->[1];
    die "malformed let*\n" if ref($bindings) ne 'MalList' and ref($bindings) ne 'MalVector';
    die "malformed let*\n" if @$bindings % 2;
    my $inner_env=$env->extend();
    while(@$bindings) {
        my $key=shift @$bindings;
        my $value=EVAL(shift @$bindings, $inner_env);
        $inner_env->set($key, $value); 
    }
    return EVAL($ast->[2], $inner_env);
}

my $line;
my $term=Term::ReadLine->new('mal');
my $prompt='user> ';
my $OUT=$term->OUT || \*STDOUT;
while(defined ($line=$term->readline($prompt))) {
    chomp $line;
    $term->addhistory($line) if $line =~ /\S/;
    eval {
        print $OUT rep($line), "\n";
    } or do {
        print $@, "\n" if $@;
    };
}

File: impls/js/step3_env.js
const readline = require('readline');
const read_str = require('./reader.js');
const pr_str = require('./printer.js');
const Env = require('./env.js');
const repl_env = new Env();

const ns = {
    '+': (a, b) => { return a + b; },
    '-': (a, b) => { return a - b; },
    '*': (a, b) => { return a * b; },
    '/': (a, b) => { return Math.trunc(a / b); },
};

for (const fn in ns) {
    repl_env.set(Symbol(fn), ns[fn]);
}

const special = {
    'def!': (form, env) => {
        const value = EVAL(form[3], env);
        env.set(form[2], value);
        return value;
    },
    'let*': (form, env) => {
        const new_env = new Env(env);
        const bindings = form[2].slice(1);
        while(bindings.length) {
            const name = bindings.shift();
            new_env.set(name, EVAL(bindings.shift(), new_env));
        }
        return EVAL(form[3], new_env);
    }
};

function READ(line) {
    return read_str(line);
}

function eval_ast(ast, env) {
    if (typeof ast === 'symbol' && !ast.description.startsWith(':')) {
        return env.get(ast);
    } else if (Array.isArray(ast) && typeof ast[0] === 'boolean') {
        return ast.map((value) => EVAL(value, env));
    } else if (typeof ast === 'object') {
        return Object.fromEntries(
            Object.entries(ast)
            .map(([key, value]) => [key, EVAL(value, env)])
        );
    } else {
        return ast;
    }
}

function EVAL(ast, env) {
    if (!Array.isArray(ast) || ast[0]) {
        return eval_ast(ast, env);
    } else if (ast.length < 2) {
        return ast;
    } else if (typeof(ast[1]) === 'symbol' && ast[1].description in special) {
        //console.log(ast);
        return special[ast[1].description](ast, env);
    }
    const evaluated_ast = eval_ast(ast, env);
    return evaluated_ast[1](...evaluated_ast.slice(2));
}
function PRINT(value) {
    return pr_str(value, true);
}

function rep(line) {
    return PRINT(EVAL(READ(line), repl_env));
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  prompt: 'user> '
});

rl.prompt();

rl.on('line', (line) => {
    try {
        if(line.trim()) {
            console.log(rep(line));
        }    
    } catch (e) {
        console.log(e);
    }
    rl.prompt();
}).on('close', () => {
  process.exit(0);
});
